libdebug/state/thread_context.py

```diff
@@ -8,115 +8,189 @@
+    def exit_signal(self: ThreadContext) -> int | None:
+        """The thread's exit signal."""
+        self._internal_debugger._ensure_process_stopped()
+        if not self.thread_dead:
+            liblog.warning("Thread is not dead. No exit signal available.")
+            return None
+        elif self._exit_signal is None and self._exit_code is not None:
+            liblog.warning("Thread exited with code %d. No exit signal available.", self._exit_code)
+            return None
+        return resolve_signal_name(self._exit_signal)
+
+    @property
+    def signal(self: ThreadContext) -> str | None:
+        """The signal will be forwarded to the thread."""
+        self._internal_debugger._ensure_process_stopped()
+        return None if self._signal_number == 0 else resolve_signal_name(self._signal_number)
+
+    @signal.setter
+    def signal(self: ThreadContext, signal: str | int) -> None:
+        """Set the signal to forward to the thread."""
+        self._internal_debugger._ensure_process_stopped()
+        if self._signal_number != 0:
+            liblog.debugger(
+                f"Overwriting signal {resolve_signal_name(self._signal_number)} with {resolve_signal_name(signal) if isinstance(signal, int) else signal}."
+            )
+        if isinstance(signal, str):
+            signal = resolve_signal_number(signal)
+        self._signal_number = signal
+        self._internal_debugger.resume_context.threads_with_signals_to_forward.append(self.thread_id)
 
     def backtrace(self: ThreadContext) -> list:
         """Returns the current backtrace of the thread."""
+        internal_debugger = self._internal_debugger
+        internal_debugger._ensure_process_stopped()
         stack_unwinder = stack_unwinding_provider()
         backtrace = stack_unwinder.unwind(self)
-        maps = self.context.debugging_interface.maps()
+        maps = internal_debugger.debugging_interface.maps()
         return [resolve_address_in_maps(x, maps) for x in backtrace]
 
     def current_return_address(self: ThreadContext) -> int:
         """Returns the return address of the current function."""
+        self._internal_debugger._ensure_process_stopped()
         stack_unwinder = stack_unwinding_provider()
         return stack_unwinder.get_return_address(self)
+
+    def step(self: ThreadContext) -> None:
+        """Executes a single instruction of the process."""
+        self._internal_debugger.step(self)
+
+    def step_until(
+        self: ThreadContext,
+        position: int | str,
+        max_steps: int = -1,
+    ) -> None:
+        """Executes instructions of the process until the specified location is reached.
+
+        Args:
+            position (int | bytes): The location to reach.
+            max_steps (int, optional): The maximum number of steps to execute. Defaults to -1.
+        """
+        self._internal_debugger.step_until(self, position, max_steps)
+
+    def finish(self: ThreadContext, exact: bool = True) -> None:
```

Comment:
The non-exact finish calls cont() after setting a breakpoint, thus affects the state of all threads, thus should not be reachable from a ThreadContext but only from the main debugger, due to its side-effects.

Reply:
Before this PR, it was possible to pass the ThreadContext as an argument (or the ThreadContext of the main thread was taken) and select the exact mode. The incongruence was exactly the same. Now, the difference is only that the thread context is not set internally or passed as an argument, but it is managed like the other functionalities. It is one of the inconsistent APIs we have, but it is, again, not strictly related to the PR.

Reply:
The incongruence comes from the fact that this is a method that affects all threads.
Previously it was callable only from the overall debugger, now it is accessible directly from the target thread.

We don't have thread.cont(), we can't have thread.finish() either until we change how the non-exact finish works, unless we make just the exact finish available from the ThreadContext

Reply:
The `finish` API will be changed soon with all other APIs. It will not have the `exact` flag anymore as it will take a string parameter that specifies the chosen heuristic. In the current draft of the documentation, the impact of the backtrace heuristic on all threads is documented. However, we may choose to resolve the inconsistency by adding a thread-specific continuation function. Still, I would say it is not a change that belongs in another pull request.